Feature pruning

cowchipkid opened the pull request:
The feature pruning implementation, prunes low value features for SVM, LTU subclasses and sparse nets.

danyaljj commented on lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/Train.java:
```diff
@@ -880,53 +880,57 @@ public void run() {
+                    
+                    // save the final model.
+                    System.out.println("Writing " + getName());
                     learner.save(); // Doesn't write .lex if lexicon is empty.
```
@cowchipkid can you clarify the change here? Confused about what's happening here ....

cowchipkid replied:
I moved the Accuracy reporting outside the block to ensure doneTraining() gets called before accuracy reporting. doneTraining will apply the feature optimization, we want the score after that. If accuracy is reported before doneTraining, we will get the score of the un-optimized models, which would be for not.

danyaljj replied:
Ah I see. Makes sense.

cowchipkid commented on lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/featurepruning/SparseNetworkOptimizer.java:
```diff
@@ -0,0 +1,158 @@
+                            continue;
+                        }
+                        double wt = ltu.getWeightVector().getRawWeights().get(fi);
+
```
Ya, this is wrong. We need to call hasWeight here I think, this won't work for SparseAveragedPerceptron where we need to sum the past average with the actual weight.

cowchipkid replied:
Also, it doesn't even address absolute zero.

cowchipkid commented:
The documentation in the feature pruning stuff is in the package-info.java in the package with all the optimization stuff. It was was already in place, I enhanced it a bit, and added some links into the life cycle methods. I also fixed one bug that likely had no impact. It would not disable pruning when you would set the threshold to zero.

danyaljj commented on lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/Lexicon.java:
```diff
@@ -742,10 +775,9 @@ public int compare(int i1, int i2) {
         for (int i = 0; i < indexes.length; ++i) {
             Feature f = inverse.get(indexes[i]);
-            previousClassName =
+             previousClassName =
```
drop the extra space?

danyaljj commented on lbjava/src/main/java/edu/illinois/cs/cogcomp/lbjava/learn/SupportVectorMachine.java:
```diff
@@ -72,6 +77,11 @@
     /** Default for {@link #bias}. */
     public static final double defaultBias = 1.0;
+    /** any weight less than this is considered irrelevant. This is for prunning. */
+    public static final double defaultFeaturePruningThreshold = 0.000001;
```
Is this always used in an absolute sense? I.e. it always have to be positive. 
In other words, if its zero, no pruning will be done.

cowchipkid replied:
@danyaljj setting the threshold to zero is how you disable pruning, effectively.

danyaljj commented:
A minor comment; Looks good to me! 
Unless @mssammon plans to go through this, feel free to merge it.

mssammon commented:
am still in the thick of post-move activity; feel free to merge.
